### PR TITLE
Simplify file upload handling

### PR DESCRIPTION
**Pull request by teohhanhui**

### CODE REVIEW

File: core/file-upload.md [hunk 1]
```diff
@@ -74,14 +74,36 @@ class MediaObject
      * @Assert\NotNull()
      * @Vich\UploadableField(mapping="media_object", fileNameProperty="contentUrl")
      */
-    public $file;
+    protected $file;
```
Comment (dunglas) [inline core/file-upload.md]: Why this change? Why try to use public properties as much as possible in docs.

Comment (teohhanhui) [reply core/file-upload.md]: Why? It's a bad practice.

Comment (dunglas) [reply core/file-upload.md]: It makes the examples simpler, it’s less verbose, and it’s not a bad practice in this case where it brings nothing: https://dev.to/scottshipp/avoid-getters-and-setters-whenever-possible-c8m

Also, getters and setters are also a bad practice... https://www.javaworld.com/javaworld/jw-09-2003/jw-0905-toolbox.html

The only benefit of getters/setters in PHP is type safety (until the typed properties RFC is accepted). And it’s not an issue with API Platform because we check the types during the deserialization.

File: core/file-upload.md [hunk 2]
```diff
@@ -97,94 +119,62 @@ that handles the file upload.
...
-        // This will be handled by API Platform and returns a validation error.
-        throw new ValidationException($this->validator->validate($mediaObject));
-    }
-}
-```
+        $mediaObject = new MediaObject();
+        $mediaObject->setFile($uploadedFile);
 
-As you can see, the action uses a form. You will need this form to be like this:
+        $this->validate($mediaObject);
```
[the remark this answers is not in the document]
Comment (jewome62) [reply core/file-upload.md]: Or integrate into validate function

[the remark this answers is not in the document]
Comment (teohhanhui) [reply core/file-upload.md]: This is using the ApiPlatform `ValidatorInterface`, so the correct `ValidationException` is already being thrown.

[the remark this answers is not in the document]
Comment (teohhanhui) [reply core/file-upload.md]: Though, to be honest, it's all kinds of weird that the `ValidatorInterface` which is not in a "Bridge" throws an exception which is in a "Bridge".

File: core/file-upload.md [hunk 3]
```diff
@@ -49,39 +49,57 @@ use ApiPlatform\Core\Annotation\ApiResource;
...
- *     "post"={
- *         "method"="POST",
- *         "path"="/media_objects",
- *         "controller"=CreateMediaObjectAction::class,
- *         "defaults"={"_api_receive"=false},
+ * @ApiResource(
+ *     iri="http://schema.org/MediaObject",
+ *     collectionOperations={
+ *         "post"={
+ *             "method"="POST",
```
Comment (dunglas) [inline core/file-upload.md]: can be removed IIRC

File: core/file-upload.md [hunk 4]
```diff
@@ -92,99 +110,115 @@ that handles the file upload.
...
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use ApiPlatform\Core\Validator\ValidatorInterface;
 use App\Entity\MediaObject;
-use App\Form\MediaObjectType;
-use Sensio\Bundle\FrameworkExtraBundle\Configuration\IsGranted;
-use Symfony\Bridge\Doctrine\RegistryInterface;
-use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
-use Symfony\Component\Form\FormFactoryInterface;
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Security;
```
Comment (dunglas) [inline core/file-upload.md]: It's not installed by default IIRC, we should explain it to the user.

Comment (teohhanhui) [reply core/file-upload.md]: Indeed, it should use API Platform's expression language instead. :)

File: core/file-upload.md [hunk 5]
```diff
@@ -92,99 +110,115 @@ that handles the file upload.
...
-        }
+        return $mediaObject;
+    }
+
+    /**
+    * @throws ValidationException
+    */
+    private function validate(MediaObject $mediaObject): void
+    {
+        $attributes = RequestAttributesExtractor::extractAttributes($request);
```
Comment (borNfreee) [inline core/file-upload.md]: `$request` is undefined

Comment (teohhanhui) [reply core/file-upload.md]: Nice catch!

File: core/file-upload.md [hunk 6]
```diff
@@ -92,99 +113,111 @@ that handles the file upload.
...
-        }
+        return $mediaObject;
+    }
 
-        // This will be handled by API Platform and returns a validation error.
-        throw new ValidationException($this->validator->validate($mediaObject));
+    /**
+    * @throws ValidationException
+    */
+    private function validate(MediaObject $mediaObject, Requst $request): void
```
Comment (Jibbarth) [inline core/file-upload.md]:
```suggestion
    private function validate(MediaObject $mediaObject, Request $request): void
```

File: core/file-upload.md [hunk 7]
```diff
@@ -92,99 +113,111 @@ that handles the file upload.
...
+        ];
     }
 
-    public function getBlockPrefix()
+    public function onPreSerialize(GetResponseForControllerResultEvent $event)
     {
-        return '';
+        $mediaObject = $event->getControllerResult();
+
+        if (!$mediaObject instanceof MediaObject) {
```
Comment (Jibbarth) [inline core/file-upload.md]: This will work only for `get item operation`, right ? Or it'll work also on `get collection operation`. 
I thought in case of collection operation, `getControllerResult` returned an `AbstractPaginator`, but may I'm wrong.

File: core/file-upload.md (hunk 6 above)
Comment (teohhanhui) [reply core/file-upload.md]: Oops! 🙈

File: core/file-upload.md (hunk 7 above)
Comment (teohhanhui) [reply core/file-upload.md]: You're right. I'll fix this.

File: core/file-upload.md [hunk 8]
```diff
@@ -49,39 +49,60 @@ use ApiPlatform\Core\Annotation\ApiResource;
...
  *     },
- * })
+ *     collectionOperations={
+ *         "post"={
+ *             "controller"=CreateMediaObjectAction::class,
+ *             "defaults"={
+ *                 "_api_receive"=false,
+ *             },
+ *             "access_control"="is_granted('ROLE_USER')",
+ *             "validation_groups"={"Default", "media_object_create"},
```
[the remark this answers is not in the document]
Comment (teohhanhui) [reply core/file-upload.md]: @Jibbarth Added and fixed the double array with `consumes`. It's supposed to be an array of strings: https://github.com/api-platform/core/blob/c80d4934ead61ffd4d5e08f3d86c79fa6fb75ce0/tests/Swagger/Serializer/DocumentationNormalizerV2Test.php#L163

[the remark this answers is not in the document]
Comment (teohhanhui) [reply core/file-upload.md]: https://swagger.io/docs/specification/2-0/file-upload/

**Comment by teohhanhui** [comment]
@dunglas Done.

**Comment by dunglas** [comment]
Thanks @teohhanhui
